### ISSUE TITLE
Make Topology Tool

### ISSUE THREAD

**Issue by LLeBlanc110** [usgs/gems-tools-arcmap#63]
![MakeTopology](https://user-images.githubusercontent.com/61474179/113889536-f2377800-9788-11eb-8d24-918e4d9615c9.JPG)

![maketopology2](https://user-images.githubusercontent.com/61474179/113889959-55290f00-9789-11eb-9397-500278bcb294.JPG)

 I'm using the most recent tools (feb. 2021), but I've tried running on the older versions and get the same result. Running in ArcCatalog produces the same result.

**Comment by LLeBlanc110** [comment]
I realized I didn't have the most recent version (1.5.5), so I updated and the same issue is still happening. Also, the validate database tool doesn't run either. 
![image](https://user-images.githubusercontent.com/61474179/113897519-91ac3900-9790-11eb-89a5-d718719f7725.png)

**Comment by LLeBlanc110** [comment]
The same errors are being thrown for the validate database and make topology tools in the pro GeMS tools (2.4.4). 

![image](https://user-images.githubusercontent.com/61474179/113901946-f8335600-9794-11eb-832e-75f305c34194.png)
![image](https://user-images.githubusercontent.com/61474179/113902092-26b13100-9795-11eb-92b0-630bdca8f923.png)

**Comment by ethoms-usgs** [comment]
Ok, I will take a look at these today.

**Comment by ethoms-usgs** [comment]
First, the Make Topology tool - I cannot recreate this error. The path to your gdb looks good and I confirmed the tool ran correctly even when the names of GeologicMap, ContactsAndFaults, and MapUnits had been changed to include a prefix. Can you include a screenshot from ArcCatalog of the contents of your gdb?

And, try running the tools from ArcCatalog completely outside of an ArcMap session. With some tools - can't confirm this is one of them without digging way into the code - having layers in the map from different data sources that have the same names can cause the tools to point to the wrong feature class.

**Comment by LindaTedrow** [comment]
Evan,
When I look at the Lainey’s error it is interestingly similar to the error with the search cursor that we “discussed” in October.

In rereading my notes from that:
# open search cursor on DMU sorted by HKey
    ### the sql clause is causing an error -- will try adding MapUnit to clause as per
    ### https://desktop.arcgis.com/en/arcmap/latest/analyze/arcpy-data-access/searchcursor-class.htm


# open search cursor on DMU sorted by HKey
    sql = (None, 'ORDER BY HierarchyKey')
    with arcpy.da.SearchCursor('DescriptionOfMapUnits',('MapUnit'),None,None,False,sql) as cursor:

From reading the ESRI Search Cursor page,  I think that the field names must include HierarchyKey in order for the sql statement to work.


Linda

Linda Tedrow
Idaho Geological Survey
University of Idaho
3rd Floor Morrill Hall
875 Perimeter Drive, MS 3014
Moscow, Id   83844-3014
(208) 885-1579

**Comment by LLeBlanc110** [comment]
@ethoms-usgs The issue likely stems from changing the names of the feature classes with a suffix, rather than a prefix. I was able to run the tools on a geodatabase where I hadn't changed the names of the feature classes. I can flip the naming around and see if I have any issues.
![image](https://user-images.githubusercontent.com/61474179/114189322-b24ecd00-990f-11eb-864a-d21ee840dd8a.png)

Also, for what it's worth I've experienced the issue you described about having ArcMap open and the tool pulling the wrong geodatabase feature classes, than the one I pointed it to.

**Comment by LLeBlanc110** [comment]
@ethoms-usgs I thought I could fix the problem by flipping the names around, since the tools allows for prefixes, but maybe not suffixes? So good news, the make topology tool works now, but I'm still getting the same error message for the validation tool. I built the geodatabase from scratch rather than using the tool, so I could've messed up a few things along the way.
![image](https://user-images.githubusercontent.com/61474179/114193529-3dca5d00-9914-11eb-9b22-e80e9c3dd2ec.png)

Upon checking, I found that I allowed Null values in the DMU for HierarchyKey, could that be the problem? 

Also, somewhat related, in building the gdb from scratch I realized the attribute field in OrientationPoints is now "StationsID" in the publication, formerly was "StationID" in the manuscript. The validate tool will flag this as an error if someone was building their gdb based on the pub. Not a big deal, just pointing it out. And lastly, for what's it worth, I've experienced the issue you describe about having ArcMap open and the validate tool pulling from another open geodatabase than the one I directed it to.

**Comment by ethoms-usgs** [comment]
The Validate Database tool does expect to be able to find a 'DescriptionOfMapUnits' table, with no prefixes or suffixes, so try removing the prefix. The Make Topology tool works with feature classes with prefixes (only -- no suffixes) because the intent is for it to be used to build topology not only in GeologicMap feature dataset, but CrossSection datasets as well, where ContactsAndFaults and MapUnitPolys do get named with prefixes. 

In general, the tools do not expect to have to deal with prefixes like what you have used because of the idea that a GeMS database describes the geology of only one map, that is, a prefix that identifies a feature class as belonging to a particular gdb is redundant. We could consider re-writing the tools to always look for names in the way that Make Topology tool does, but the workaround for now is to not rename those required table names

**Comment by LLeBlanc110** [comment]
Changing the DMU back to its original name fixed the issue and the tool ran. I understand the need to keep the names as is, I did it for clarity since I'm compiling 4 GeMS geodatabases in one map. Changing the names is easy enough, and I should've just changed the alias (_duh_). Thanks for the help, Evan!

**Comment by ethoms-usgs** [comment]
Yeah, I wondered if you were doing something like that!
Good to know.

**Comment by ethoms-usgs** [comment]
@LindaTedrow Yes! It is similar isn't it? I am still curious about the error you reported. I have continued the discussion over at the original [issue](https://github.com/usgs/gems-tools-arcmap/issues/49)
